extensions/amp-megaphone/0.1/amp-megaphone.js

```diff
@@ -0,0 +1,234 @@
+  constructor(element) {
+    super(element);
+
+    /** @private {?Element} */
+    this.iframe_ = null;
+
+    /** @private {?Function} */
+    this.unlistenMessage_ = null;
+
+    /** @private {boolean} */
+    this.isPlaylist_ = false;
+
+    /** @private {string} */
+    this.baseUrl_ = '';
+  }
+
+  /**
+   * @param {boolean=} opt_onLayout
+   * @override
+   */
+  preconnectCallback(opt_onLayout) {
+    this.preconnect.url(this.baseUrl_, opt_onLayout);
+    this.preconnect.url('https://assets.megaphone.fm', opt_onLayout);
+    this.preconnect.url('https://megaphone.imgix.net', opt_onLayout);
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return (
+      layout == Layout.FIXED ||
+      layout == Layout.FIXED_HEIGHT ||
+      layout == Layout.FILL ||
+      layout == Layout.NODISPLAY
+    );
+  }
+
+  /** @override */
+  buildCallback() {
+    this.updateBaseUrl_();
+  }
+
+  /** @override */
+  mutatedAttributesCallback(mutations) {
+    const playlist = mutations['data-playlist'];
+    const episode = mutations['data-playlist'];
+    if (playlist !== undefined || episode !== undefined) {
+      this.updateBaseUrl_();
+    }
+
+    const light = mutations['data-light'];
+    const sharing = mutations['data-sharing'];
+    const episodes = mutations['data-episodes'];
+    const start = mutations['data-start'];
+    const tile = mutations['data-tile'];
+    if (
+      light !== undefined ||
+      sharing !== undefined ||
+      episodes !== undefined ||
+      start !== undefined ||
+      tile !== undefined
+    ) {
+      if (this.iframe_) {
+        this.iframe_.src = this.getIframeSrc_();
+      }
+    }
+  }
+
+  /**@override*/
+  layoutCallback() {
+    const height = this.element.getAttribute('height');
+    const iframe = this.element.ownerDocument.createElement('iframe');
+
+    iframe.setAttribute('frameborder', 'no');
+    iframe.setAttribute('scrolling', 'no');
+
+    iframe.src = this.getIframeSrc_();
+
+    this.unlistenMessage_ = listen(
+      this.win,
+      'message',
+      this.handleMegaphoneMessages_.bind(this)
+    );
+
+    this.applyFillContent(iframe);
+    iframe.height = height;
+    this.element.appendChild(iframe);
+
+    this.iframe_ = iframe;
+
+    return this.loadPromise(iframe);
+  }
+
+  /** @override */
+  unlayoutCallback() {
+    if (this.iframe_) {
+      removeElement(this.iframe_);
+      this.iframe_ = null;
+    }
+    if (this.unlistenMessage_) {
+      this.unlistenMessage_();
+    }
+    return true; // Call layoutCallback again.
+  }
+
+  /**
+   * @return {string}
+   * @private
+   */
+  getIframeSrc_() {
+    this.updateBaseUrl_();
+    let url = this.baseUrl_ + '/';
+    const mediaid = userAssert(
+      this.element.getAttribute('data-playlist') ||
+        this.element.getAttribute('data-episode'),
+      'data-playlist or data-episode is required for <amp-megaphone> %s',
+      this.element
+    );
+
+    if (!this.isPlaylist_) {
```

Comment:
Let's start using the URL construction facilities where we can here.

A simple replacement here is also prudent.
`const url = this.baseUrl_ + !this.isPlaylist_ ? '/mediaid/' : '/';`